include/constantine/protocols/ethereum_verkle_ipa.h

```diff
@@ -0,0 +1,228 @@
+/** Constantine
+ *  Copyright (c) 2018-2019    Status Research & Development GmbH
+ *  Copyright (c) 2020-Present Mamy André-Ratsimbazafy
+ *  Licensed and distributed under either of
+ *    * MIT license (license terms in the root directory or at http://opensource.org/licenses/MIT).
+ *    * Apache v2 license (license terms in the root directory or at http://www.apache.org/licenses/LICENSE-2.0).
+ *  at your option. This file may not be copied, modified, or distributed except according to those terms.
+ */
+#ifndef __CTT_H_ETHEREUM_VERKLE_IPA__
+#define __CTT_H_ETHEREUM_VERKLE_IPA__
+
+#include "constantine/core/datatypes.h"
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef enum __attribute__((__packed__)) {
+    cttEthVerkleIpa_Success,
+    cttEthVerkleIpa_VerificationFailure,
+    cttEthVerkleIpa_InputsLengthsMismatch,
+    cttEthVerkleIpa_ScalarZero,
+    cttEthVerkleIpa_ScalarLargerThanCurveOrder,
+    cttEthVerkleIpa_EccInvalidEncoding,
+    cttEthVerkleIpa_EccCoordinateGreaterThanOrEqualModulus,
+    cttEthVerkleIpa_EccPointNotOnCurve,
+    cttEthVerkleIpa_EccPointNotInSubGroup
+} ctt_eth_verkle_ipa_status;
+
+static const char* ctt_eth_verkle_ipa_status_to_string(ctt_eth_verkle_ipa_status status) {
+  static const char* const statuses[] = {
+    "cttEthVerkleIpa_Success",
+    "cttEthVerkleIpa_VerificationFailure",
+    "cttEthVerkleIpa_InputsLengthsMismatch",
+    "cttEthVerkleIpa_ScalarZero",
+    "cttEthVerkleIpa_ScalarLargerThanCurveOrder",
+    "cttEthVerkleIpa_EccInvalidEncoding",
+    "cttEthVerkleIpa_EccCoordinateGreaterThanOrEqualModulus",
+    "cttEthVerkleIpa_EccPointNotOnCurve",
+    "cttEthVerkleIpa_EccPointNotInSubGroup"
+  };
+  size_t length = sizeof statuses / sizeof *statuses;
+  if (0 <= status && status < length) {
+    return statuses[status];
+  }
+  return "cttEthVerkleIpa_InvalidStatusCode";
+}
+
+// Opaque types for Nim-defined types
+typedef struct Fr Fr;
+typedef struct Banderwagon Banderwagon;
+typedef struct EC_TwEdw EC_TwEdw;
+typedef struct Fp Fp;
```

Comment:
All the fields and curves related struct should be defined in `constantine/include/constantine/curves/banderwagon.h`

This would be similar to pallas and vesta as they don't use pairings: https://github.com/mratsim/constantine/blob/2c3de8738d031394dfb78e3fc839cf5d12e42b92/bindings/lib_curves.nim#L68-L100

However Banderwagon is an Edwards curve so the template there might need an adaptation.
I don't remember what I add in mind as differences between short-weierstrass and twisted edwards, maybe it was just the affine<->jacobian<->projective coordinate conversion and that could be done as a separate template, see https://github.com/mratsim/constantine/blob/2c3de8738d031394dfb78e3fc839cf5d12e42b92/bindings/c_curve_decls.nim#L314-L432

Reply:
@mratsim  i resolved this and added separate header for banderwagon. Please review and suggest changes. Also how can i verify things are working as expected?

Reply:
Good question, I don't have a testsuite for the C API of elliptic curves. It's not ideal but for now we can skip it merge as-is and focus on the C API for the full protocol.